escape ' -> \' for the UI-WELCOME_ERROR

srl295 opened the pull request:
* add a new function i18n_singleQuoteVar() for ui_app.jade so that “err.innerText = '#{i18n_singleQuoteVar('UI-WELCOME_ERROR')}';” doesn’t fail if the UI-WELCOME_ERROR contains a single quote

Fixes: https://github.com/elastic/kibana/issues/10580

elasticmachine commented:
Can one of the admins verify this patch?

srl295 commented:
FYI @hickeyma

epixa commented:
jenkins, test this

hickeyma commented on src/ui/views/ui_app.jade:
```diff
@@ -125,7 +125,7 @@ block content
           err.style['padding'] = '25px';
-          err.innerText = '#{i18n('UI-WELCOME_ERROR')}';
+          err.innerText = '#{i18n_singleQuoteVar('UI-WELCOME_ERROR')}';
 
```
I am wondering would it be better here to assign to double quotes instead to escape the single quote:
err.innerText = **"**#{i18n('UI-WELCOME_ERROR')}**"**;

srl295 replied:
@hickeyma I tried that, actually using `JSON.stringify` instead and double quotes… but Jade (now pug) ends up quoting oddly- I think a html safe transform is implicated.

srl295 replied:
still would need to quote doublequotes in that case.

srl295 replied:
for that matter, this line could just be:
>`err.innerText = '#{i18n('UI-WELCOME_ERROR').replace(/'/g,'\\\'')}';

and drop the other function.

hickeyma replied:
I tried this and it works as expected without any other cahnges:
err.innerText = "#{i18n('UI-WELCOME_ERROR')}";

srl295 replied:
will that work if a `"` shows up in the string?

srl295 replied:
what about something like `err.innerText = "#{i18n('UI-WELCOME_ERROR').replace(/"/g,'\\"')}";` for this one spot (if we don't use the meta tag option)

_you are in a maze of twisty nested quoted strings, some alike_

hickeyma commented on src/ui/index.js:
```diff
@@ -101,7 +101,10 @@ export default async (kbnServer, server, config) => {
         i18n: key => _.get(translations, key, ''),
+        // var v = #{i18n_var('KEY')} -> var v = "value (\"Hi!\")"
+        i18n_singleQuoteVar: key => _.get(translations, key, '').replace(/'/g,'\\\''),
       });
```
Not sure we would want 2 variables passed which contain the separate instances of the translations. Refer to comment in  'src/ui/views/ui_app.jade'  by using double quotes.

srl295 replied:
they're two functions… and actually, they are only for this startup page.

hickeyma replied:
Ahhhh ..  meant functions!  Just trying to avoid having to make changes if another means possible.

spalger commented:
Take notice how we handle this here: https://github.com/srl295/kibana/blob/559bb8bf3cbad7a189681f9ae864fc4f9f1cbe0c/src/ui/views/ui_app.jade#L104

srl295 commented:
@spalger I thought about that, but was not sure how to stringify the existing function. `!{JSON.stringify(#{i18n(…)})}` I didn't think would work becasue `#{` would be interpreted in JS, but perhaps that's not true.

spalger commented:
@srl295 the JSON stringification happens server side, so the line would look like:

```js
err.innerText = !{JSON.stringify(i18n(…))};
```

Now that I think about it, that wouldn't work if the string had html in it 😞 . Escaping in a script tag is hard (which is why we ended up using a meta tag for our injected vars).

epixa commented:
Can we just use a meta tag for this?

srl295 commented:
@epixa @spalger yeah— we could stash the message off somewhere… and then go find it later. Then Jade^H^H^H^Hpug handles the html safe escaping.

spalger commented:
@hickeyma it would probably work, but I'm just not comfortable writing an escaping function inline (or elsewhere) unless it's tested and the requirements of the function are better researched/more certain.

srl295 commented:
@spalger let me try the meta tag option.

srl295 commented:
@epixa @spalger @hickeyma hey, enjoy your conf! And FYI, new commit using a div (keeps the messages together)

srl295 commented:
I've rebased now that #8766 is merged.

epixa commented:
jenkins, test this

spalger commented:
jenkins, test this

tsullivan commented:
looks like there is a conflict in this PR. Does `src/ui/views/ui_app.jade` need to be `src/ui/ui_render/views/ui_app.jade` ?

> LGTM. @spalger and @Bargs, what do you think?

Ping  @spalger and @Bargs

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/129/)

srl295 commented:
Hey happy new year and all!  I think this may need some serious rebasing.

srl295 commented:
Old commit was https://github.com/srl295/kibana/commit/c17baa493340afbe9318109d8733310223ddcb7b — pushing a rebase now.

Bargs commented:
Not sure if this is still relevant or not, Spencer would probably know best.

srl295 commented:
I couldn't remember how I tested this. I'm doing this locally to test:

```diff
diff --git a/src/core_plugins/kibana/translations/en.json b/src/core_plugins/kibana/translations/en.json
index ddb7a272e..37e7553a2 100644
--- a/src/core_plugins/kibana/translations/en.json
+++ b/src/core_plugins/kibana/translations/en.json
@@ -1,4 +1,4 @@
 {
   "UI-WELCOME_MESSAGE": "Loading Kibana",
-  "UI-WELCOME_ERROR": "Kibana did not load properly. Check the server output for more information."
+  "UI-WELCOME_ERROR": "Kibana didn't load properly. Check the server output for more information."
 }
diff --git a/src/ui/ui_render/bootstrap/template.js.hbs b/src/ui/ui_render/bootstrap/template.js.hbs
index 8a4c5fe52..6ed394f6e 100644
--- a/src/ui/ui_render/bootstrap/template.js.hbs
+++ b/src/ui/ui_render/bootstrap/template.js.hbs
@@ -40,6 +40,8 @@ window.onload = function () {
     dom.setAttribute('async', '');
     dom.addEventListener('error', failure);

+    failure();
+
     if (type === 'script') {
       dom.setAttribute('src', file);
       dom.addEventListener('load', next);
```

srl295 commented:
@Bargs Hi! I was able to get a rebase to work. I'm checking to see if I can repro the original issue.

srl295 commented:
Yup I can repro. 

see my above comment, fails (spins forever on loading) because of:

> `      err.innerText = 'Kibana didn't load properly. Check the server output for more information.';`

So, it's still relevant. And the PR fixes it. Rerun it through CI?

Bargs commented:
jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/146/)

Bargs commented:
I didn't test the fix but the code looks ok to me. I'm going to assign @azasypkin as a reviewer as he's taking over the internationalization effort, so he can decide if it makes sense to merge this or not.

azasypkin commented on src/ui/ui_render/views/ui_app.jade:
```diff
@@ -109,5 +113,7 @@ block content
         .kibanaWelcomeLogo
     .kibanaWelcomeText
       | #{i18n('UI-WELCOME_MESSAGE')}
+    #kibanaLoader__error
```
__suggestion:__ what if we use data attribute instead?
```jade
.kibanaWelcomeText(data-error-message='#{i18n("UI-WELCOME_ERROR")}')
```

It won't require additional style with `display: none`, in js we can extract it like this:
```js
document.querySelector('[data-error-message]').dataset.errorMessage
// or `.getAttribute('data-error-message')` if we want to support IE 10 - trying to figure that out at the moment).
```

Or even maybe add this attribute to body in `chrome.jade` and then we won't need `querySelector` at all.

What do you think?

srl295 replied:
@azasypkin makes some sense.  But still uses the i18n helper?

azasypkin replied:
> But still uses the i18n helper?

Yes, but see my comment above, we need Jade `i18n` helper, not the one we used in `Handlebars` template.

azasypkin replied:
Sorry, it's confusing, I know :)

srl295 commented on src/ui/ui_render/bootstrap/template.js.hbs:
```diff
@@ -22,13 +22,15 @@ window.onload = function () {
       err.style['background'] = '#F44336';
       err.style['padding'] = '25px';
-      err.innerText = '{{i18n 'UI-WELCOME_ERROR'}}';
+      err.innerText = failureText;
```
@azasypkin hm - well, the point here is to load the localized error. What's the right way to do it now?

The entire bug here is that at runtime,

```js
err.innerText = '{{i18n 'UI-WELCOME_ERROR'}}';
```

could (with the right translation) get replaced by the i18n helper to:

```js
err.innerText = 'Quel temps va-t-il faire aujourd'hui?';
```

… which is invalid JS.

azasypkin replied:
Right, I mean previously we rendered message with the help of `i18n` __Handlebars__ helper (see [here](https://github.com/elastic/kibana/blob/master/src/ui/ui_render/bootstrap/app_bootstrap.js#L19)) that led to invalid JS. In this PR you're extracting error message from the DOM, so we don't need this helper anymore.

`i18n` __Jade__ helper used in `ui_app.jade` or `chrome.jade` (see [here](https://github.com/elastic/kibana/blob/master/src/ui/ui_render/ui_render_mixin.js#L132) is a different thing, we still need it.

srl295 replied:
got it. now when i get a chance I'll rebase and update.

srl295 replied:
@azasypkin Done, please re review?

markov00 commented:
Hey @srl295 Thanks for your contribution, the i18n implementation has changed in the last months, do you have time to rebase to master.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/4531/)

srl295 commented:
@markov00 I have addressed @azasypkin ’s suggestions and rebased.  However, I can't figure out how to 'break' Kibana in order to get the error message to show up now. Do you have any suggestions?

markov00 commented:
Jenkins, test this

@srl295  thanks for that 👍 
@azasypkin have you any  suggestion on how to test this? Do we already have testing suite for this cases?

azasypkin commented on src/ui/ui_render/views/ui_app.pug:
```diff
@@ -107,7 +107,7 @@ block content
       .kibanaWelcomeLogoCircle
         .kibanaWelcomeLogo
-    .kibanaWelcomeText
+    .kibanaWelcomeText(data-error-message='#{i18n("UI-WELCOME_ERROR")}')
```
__issue:__ that's why CI is failing (https://kibana-ci.elastic.co/job/elastic+kibana+pull-request+multijob-intake/4504/console):
```
12:49:04 [10:49:04] →  I18N ERROR  Error in src/ui/ui_render/views/ui_app.pug
12:49:04 Error: Empty defaultMessage in i18n() or i18n.translate() is not allowed ("UI-WELCOME_ERROR").
12:49:04 ERROR  I18N ERROR  Error in src/ui/ui_render/views/ui_app.pug
12:49:04       Error: Empty defaultMessage in i18n() or i18n.translate() is not allowed ("UI-WELCOME_ERROR").
```

I guess you wanted to write something like this instead
```pug
.kibanaWelcomeText(data-error-message="#{i18n('common.ui.welcomeError', { defaultMessage: 'Kibana did not load properly. Check the server output for more information.' })}")`
```

Our tools now enforce default message for all labels that allows developers to actually see the real label content without looking into translation files. So we got rid of `en.json` entirely and extract all default messages automatically whenever translation vendors need them.

@pavel06081991 @maryia-lapata can you please help @srl295 to test this change (i.e where one should put and register test translation file for  `common.ui.` etc.)? Essentially we need a localized version of this label that includes apostrophe.

maryia-lapata replied:
@srl295 I tested this
```
.kibanaWelcomeText(data-error-message='#{i18n("common.ui.welcomeErrorMessage", { defaultMessage: "Kibana did not load properly. Check the server output for more information." })}')
```
with `i18n` tool. It works, the message is extracted to `en.json` by the tool (@azasypkin `common.ui` namespace is already in the [config](https://github.com/elastic/kibana/blob/master/.i18nrc.json#L3)).

azasypkin replied:
Hey @srl295! Please let us know if you need any help with this PR. We intend to ship fix for this issue in the 6.6 release, but If you don't have time that's fine too! Someone from the i18n team will just take it over.

azasypkin commented:
Hey @srl295, it looks like you were busy with other stuff lately, but no worries! Since we wanted to have the fix for this issue in 6.7 we moved forward and fixed it in https://github.com/elastic/kibana/pull/29581 (basically a bit cleaned up version of your PR). 

Thank you for your contribution!
